### DISCUSSION TITLE
Scan with .xctestplan only works when tests are *not* run in parallel?

### DISCUSSION THREAD

**Discussion by ChristopherDrum** [fastlane/fastlane#19012] [Q&A]
Having a very hard time getting Scan working with fastlane 2.186.0
Specifically, I've been trying to get reliable execution on CircleCI,
but as I dug in further I realized I'm not even able to get reliable results on my local machine.

I have an empty, dummy app I'm using as a test bed.
It uses an .xctestplan (to mimic the behavior of my real app) which contains only 2 default tests that come with a new .xctestplan file. I have them doing an `assert(true)` as the "work" for the test. Absolutely trivial stuff.

I keep encountering this message of success, with no actual tests run. 
I expect to see "2" tests run with "0" failures.
```
[14:56:27]: ▸ Test Succeeded
+--------------------+---+
|      Test Results      |
+--------------------+---+
| Number of tests    | 0 |
| Number of failures | 0 |
+--------------------+---+
```

I just discovered that if I turn off "Execute in parallel" (shown here)
![Screen Shot 2021-06-28 at 2 58 35 PM](https://user-images.githubusercontent.com/320377/123587485-549c9380-d821-11eb-947b-c6b6921a4b8d.png)

I then get the expected result
```
[14:53:20]: ▸ Test Suite SNDummyTests.xctest started
[14:53:20]: ▸ SNDummyTests
[14:53:20]: ▸     ✓ testExample (0.002 seconds)
[14:53:21]: ▸     ◷ testPerformanceExample measured (0.000 seconds)
[14:53:21]: ▸     ✓ testPerformanceExample (0.573 seconds)
[14:53:21]: ▸ 	 Executed 2 tests, with 0 failures (0 unexpected) in 0.575 (0.577) seconds
[14:53:21]: ▸ 
[14:53:21]: ▸ 2021-06-28 14:53:21.387 xcodebuild[77724:3364683] [MT] IDETestOperationsObserverDebug: 19.562 elapsed -- Testing started completed.
[14:53:21]: ▸ 2021-06-28 14:53:21.387 xcodebuild[77724:3364683] [MT] IDETestOperationsObserverDebug: 0.000 sec, +0.000 sec -- start
[14:53:21]: ▸ 2021-06-28 14:53:21.387 xcodebuild[77724:3364683] [MT] IDETestOperationsObserverDebug: 19.562 sec, +19.562 sec -- end
[14:53:21]: ▸ Test Succeeded
```

My research today didn't turn up any mention of this issue, but I'm also quite frustrated spending an entire day tracking down a single checkbox, so maybe I missed something.

Anyone else encountering this? Is this a known issue?

UPDATE:
This got tests working on my local machine; no dice on CircleCI though.
CircleCI-related issue is moved into a new discussion.

**Comment by ChristopherDrum** [comment]
Update: this actually did fix the CircleCI problem. I was just dumb and committed the wrong testplan file to the repo 😑 
So the original statement applies: parallel testing doesn't work with Scan

**Comment by benlaverriere** [comment]
I've experienced this as well, although in my case there _was_ accurate output on the command line:

```
** TEST FAILED **
[18:37:50]: Exit status: 65
+-------------------------+---------------------+
|                 Test Results                  |
+-------------------------+---------------------+
| Number of tests         | 252 (and 3 retries) |
| Number of tests skipped | 2                   |
| Number of failures      | 1                   |
+-------------------------+---------------------+
```

The HTML and JUnit XML reports, however, showed zero tests (HTML) or less (XML — the file just contained one line, `<?xml version='1.0' encoding='UTF-8'?>`). As with @ChristopherDrum's case, disabling parallelization on the test plan produced correct/detailed reports. This is on fastlane 2.210.1.

I don't know much about how test reporting works under the hood, before `scan` summarizes things and produces reports. I can try to poke around but I'd love to hear from someone who knows more about the inner workings. 🙂

**Reply by benlaverriere** [reply]
Haven't dug into Fastlane code yet, but https://github.com/benlaverriere/fastlane-scan-parallel-testplans is a minimal repro case (and what I'll use to test my understanding / potential fixes)

**Reply by benlaverriere** [reply]
Aha! This comes down to differences between `xcpretty` and the newly-recommended `xcbeautify`. See [here](https://docs.fastlane.tools/best-practices/xcodebuild-formatters/#:~:text=Supports%20Xcode%27s%20parallel%20testing%20output) in the docs: `xcbeautify` supports parsing parallel testing output, but `xcpretty` does not. For my purposes, switching to `xcbeautify` means losing an HTML test report (which I don't need) but gaining the ability to run tests in parallel.

I'll see if there's a good place to mention this in `scan`'s docs more directly, since I imagine many projects are moving towards test plans these days. Ideally `scan` would detect a test plan with parallelization and warn about using `xcpretty` with it, but that might be beyond what I can contribute right now.

**Reply by benlaverriere** [reply]
Proposing some auto-detection of this case in https://github.com/fastlane/fastlane/discussions/20775
